api/envoy/extensions/filters/network/geoip/v3/geoip.proto

```diff
@@ -37,4 +38,31 @@ message Geoip {
   // Geoip driver specific configuration which depends on the driver being instantiated.
   // [#extension-category: envoy.geoip_providers]
   config.core.v3.TypedExtensionConfig provider = 2 [(validate.rules).message = {required: true}];
+
+  // Configuration for dynamically extracting the client IP address used for geolocation lookups.
+  //
+  // This field accepts a :ref:`SubstitutionFormatString <envoy_v3_api_msg_config.core.v3.SubstitutionFormatString>`
+  // that is evaluated at connection time to produce the client IP address.
+  // The formatted result must be a valid IPv4 or IPv6 address string.
+  //
+  // This allows flexible extraction of the client IP using any available
+  // :ref:`command operator <config_access_log_command_operators>`. For example:
+  //
+  // * ``%FILTER_STATE(my.custom.client.ip:PLAIN)%`` - Read from filter state populated by a preceding filter.
+  // * ``%DYNAMIC_METADATA(namespace:key)%`` - Read from dynamic metadata.
+  //
+  // If not set, the filter uses the downstream connection's remote address.
+  //
+  // If set but the formatted result is empty or not a valid IP address, the filter
+  // falls back to using the downstream connection source address.
+  //
+  // Example configuration using filter state:
+  //
+  // .. code-block:: yaml
+  //
+  //   client_ip_config:
+  //     text_format_source:
+  //       inline_string: "%FILTER_STATE(my.custom.client.ip:PLAIN)%"
+  //
+  config.core.v3.SubstitutionFormatString client_ip_config = 3;
```

Comment:
I don't want to be picky. But see, we expect a IP from this format, then, if the what will happens if the SubstitutionFormatString.json_format is used?

I mean, if SubstitutionFormatString is used here, most of fields of SubstitutionFormatString will make no sense for users of geoip and bring only confusion/chaos.

I will suggest a single string here as new field (or a simple wrapper to a single string field), like what we do in the header value mutation/custom tag/rate limit and so on.